### PR TITLE
[Refactor]: bbox coder support batch infer

### PR DESCRIPTION
**Pull request by hhaAndroid**
box coder support batch infer

### CODE REVIEW

File: mmdet/core/bbox/coder/delta_xywh_bbox_coder.py [hunk 1]
```diff
@@ -70,6 +71,8 @@ def decode(self,
         """
 
         assert pred_bboxes.size(0) == bboxes.size(0)
+        if pred_bboxes.ndim == 3:
+            assert pred_bboxes.size(1) == bboxes.size(1)
```
Comment (ZwwWayne) [inline mmdet/core/bbox/coder/delta_xywh_bbox_coder.py:75]: Should also test size(0)?

File: mmdet/core/bbox/coder/delta_xywh_bbox_coder.py [hunk 2]
```diff
@@ -198,28 +207,24 @@ def delta2bbox(rois,
...
-            y2 = y2.clamp(min=0, max=max_shape[0])
-    bboxes = torch.stack([x1, y1, x2, y2], dim=-1).view(deltas.size())
+            if isinstance(max_shape, list):
+                assert len(max_shape) == x1.shape[0]
+            max_shape = x1.new_tensor(max_shape)[..., :2]
+
+        min_xy = x1.new_tensor(0)
+        max_xy = torch.cat(
+            [max_shape] * (deltas.size(-1) // 2),
+            dim=-1).flip(dims=[-1]).unsqueeze(-2)
```
Comment (jshilong) [inline mmdet/core/bbox/coder/delta_xywh_bbox_coder.py:226]: `flip(-1)` would be ok

File: mmdet/core/bbox/coder/delta_xywh_bbox_coder.py [hunk 3]
```diff
@@ -137,21 +140,23 @@ def delta2bbox(rois,
...
+            Maximum bounds for boxes. specifies (H, W, C) or (H, W)
         wh_ratio_clip (float): Maximum aspect ratio for boxes.
         clip_border (bool, optional): Whether clip the objects outside the
             border of the image. Defaults to True.
 
     Returns:
-        Tensor: Boxes with shape (N, 4), where columns represent
-            tl_x, tl_y, br_x, br_y.
+        Tensor: Boxes with shape (B, N, 4 * num_classes) or (B, N, 4) or
+           (B, N, 4) or (N, 4), where 4 represent tl_x, tl_y, br_x, br_y.
```
Comment (jshilong) [inline mmdet/core/bbox/coder/delta_xywh_bbox_coder.py:159]: `(B, N, 4)`  repeated

File: mmdet/core/bbox/coder/tblr_bbox_coder.py [hunk 4]
```diff
@@ -133,43 +133,57 @@ def tblr2bboxes(priors,
...
+        if isinstance(max_shape, torch.Tensor):
+            max_shape = max_shape[..., :2].type_as(priors)
+        else:
+            if isinstance(max_shape, list):
+                assert len(max_shape) == priors.shape[0]
+            max_shape = priors.new_tensor(max_shape)[..., :2]
+
+        min_xy = priors.new_tensor(0)
+        max_xy = torch.cat([max_shape, max_shape],
+                           dim=-1).flip(dims=[-1]).unsqueeze(-2)
```
Comment (jshilong) [inline mmdet/core/bbox/coder/tblr_bbox_coder.py:185]: `flip(-1)`

File: mmdet/core/bbox/coder/tblr_bbox_coder.py [hunk 5]
```diff
@@ -133,43 +133,57 @@ def tblr2bboxes(priors,
...
-    xmin = prior_centers[:, 0].unsqueeze(1) - left
-    xmax = prior_centers[:, 0].unsqueeze(1) + right
-    ymin = prior_centers[:, 1].unsqueeze(1) - top
-    ymax = prior_centers[:, 1].unsqueeze(1) + bottom
-    boxes = torch.cat((xmin, ymin, xmax, ymax), dim=1)
+        wh = priors[..., 2:4] - priors[..., 0:2]
+        w, h = torch.split(wh, 1, dim=-1)
+        loc_decode[..., :2] *= h  # tb
+        loc_decode[..., 2:] *= w  # lr
+    top, bottom, left, right = loc_decode.split((1, 1, 1, 1), dim=-1)
```
Comment (jshilong) [inline mmdet/core/bbox/coder/tblr_bbox_coder.py:167]: `loc_decode.split(1, dim=-1)` may be better.

File: mmdet/core/bbox/coder/tblr_bbox_coder.py [hunk 6]
```diff
@@ -133,43 +133,57 @@ def tblr2bboxes(priors,
...
+
     if clip_border and max_shape is not None:
-        boxes[:, 0].clamp_(min=0, max=max_shape[1])
-        boxes[:, 1].clamp_(min=0, max=max_shape[0])
-        boxes[:, 2].clamp_(min=0, max=max_shape[1])
-        boxes[:, 3].clamp_(min=0, max=max_shape[0])
-    return boxes
+        if isinstance(max_shape, torch.Tensor):
+            max_shape = max_shape[..., :2].type_as(priors)
+        else:
```
Comment (jshilong) [inline mmdet/core/bbox/coder/tblr_bbox_coder.py:178]: misses the case of `tuple[int]`

File: mmdet/core/bbox/coder/delta_xywh_bbox_coder.py (hunk 1 above)
Comment (hhaAndroid) [reply mmdet/core/bbox/coder/delta_xywh_bbox_coder.py:75]: The shape may be (B, N, 4) or (N, 4), so I think it might be asserted.

File: mmdet/core/bbox/coder/tblr_bbox_coder.py (hunk 6 above)
Comment (hhaAndroid) [reply mmdet/core/bbox/coder/tblr_bbox_coder.py:178]: if type is tuple[int]，`max_shape = priors.new_tensor(max_shape)[..., :2]`

File: mmdet/core/bbox/coder/tblr_bbox_coder.py [hunk 7]
```diff
@@ -133,43 +134,57 @@ def tblr2bboxes(priors,
...
+            if isinstance(max_shape, list):
```
Comment (ZwwWayne) [inline mmdet/core/bbox/coder/tblr_bbox_coder.py:180]: may use sequential, or (list, tuple)

File: mmdet/core/bbox/coder/tblr_bbox_coder.py [hunk 8]
```diff
@@ -133,43 +134,57 @@ def tblr2bboxes(priors,
...
+
+    bboxes = torch.cat((xmin, ymin, xmax, ymax), dim=-1)
+
     if clip_border and max_shape is not None:
-        boxes[:, 0].clamp_(min=0, max=max_shape[1])
-        boxes[:, 1].clamp_(min=0, max=max_shape[0])
-        boxes[:, 2].clamp_(min=0, max=max_shape[1])
-        boxes[:, 3].clamp_(min=0, max=max_shape[0])
-    return boxes
+        if isinstance(max_shape, torch.Tensor):
```
Comment (ZwwWayne) [inline mmdet/core/bbox/coder/tblr_bbox_coder.py:177]: Can we simplify the logic here, e.g.,:
```python
if not isinstance(max_shape, torch.Tensor):
    max_shape = priors.new_tensor(max_shape)
max_shape = max_shape[..., :2].type_as(priors)

```

File: tests/test_utils/test_coder.py [hunk 9]
```diff
@@ -19,3 +21,64 @@ def test_yolo_bbox_coder():
          [41.2068, -8.9232, 181.4236, 48.5840]])
     assert expected_decode_bboxes.allclose(
         coder.decode(bboxes, pred_bboxes, grid_size))
+
+
+def test_delta_bbox_coder():
```
Comment (jshilong) [inline tests/test_utils/test_coder.py:26]: It is better to test boxes that are shaped as `N ×0 × 4` and `0 * 4`, I remember that the previous version is supported this shape (not very sure), and it’s best to test it and make sure this version has same behaviors.

File: mmdet/core/bbox/coder/tblr_bbox_coder.py [hunk 10]
```diff
@@ -133,43 +135,58 @@ def tblr2bboxes(priors,
...
           encoded boxes. By list, it represents the normalization factors at
           tblr dims. By float, it is the unified normalization factor at all
           dims. Default: 4.0
         normalize_by_wh (bool): Whether the tblr coordinates have been
           normalized by the side length (wh) of prior bboxes.
-        max_shape (tuple, optional): Shape of the image. Decoded bboxes
-          exceeding which will be clamped.
+        max_shape (Sequence[int] or torch.Tensor or Sequence[
+            Sequence[int]],optional): Maximum bounds for boxes, specifies
+            (H, W, C) or (H, W). If priors shape is (B, N, 4),then the
```
Comment (jshilong) [inline mmdet/core/bbox/coder/tblr_bbox_coder.py:149]: If priors shape is (B, N, 4), then the max_shape should be a Sequence[Sequence[int]] and the length of max_shape should also be B.

File: mmdet/core/bbox/coder/tblr_bbox_coder.py [hunk 11]
```diff
@@ -133,43 +135,58 @@ def tblr2bboxes(priors,
...
+            length of max_shape should also be B.
         clip_border (bool, optional): Whether clip the objects outside the
             border of the image. Defaults to True.
 
     Return:
-        encoded boxes (Tensor), Shape: (n, 4)
+        encoded boxes (Tensor), Shape: (N, 4) or (B, N, 4)
```
Comment (jshilong) [inline mmdet/core/bbox/coder/tblr_bbox_coder.py:155]: encoded boxes (Tensor): Shape: (N, 4) or (B, N, 4)

File: mmdet/core/bbox/coder/tblr_bbox_coder.py [hunk 12]
```diff
@@ -133,43 +135,58 @@ def tblr2bboxes(priors,
...
     if clip_border and max_shape is not None:
-        boxes[:, 0].clamp_(min=0, max=max_shape[1])
-        boxes[:, 1].clamp_(min=0, max=max_shape[0])
-        boxes[:, 2].clamp_(min=0, max=max_shape[1])
-        boxes[:, 3].clamp_(min=0, max=max_shape[0])
-    return boxes
+        if not isinstance(max_shape, torch.Tensor):
+            max_shape = priors.new_tensor(max_shape)
+        max_shape = max_shape[..., :2].type_as(priors)
+        if max_shape.ndim == 2:
```
Comment (jshilong) [inline mmdet/core/bbox/coder/tblr_bbox_coder.py:183]: add 
`assert  bboxes.dim == 3`

File: mmdet/core/bbox/coder/delta_xywh_bbox_coder.py [hunk 13]
```diff
@@ -137,21 +142,26 @@ def delta2bbox(rois,
...
+            If rois shape is (B, N, 4),then the length of max_shape should
+            also be B.
         wh_ratio_clip (float): Maximum aspect ratio for boxes.
         clip_border (bool, optional): Whether clip the objects outside the
             border of the image. Defaults to True.
 
     Returns:
-        Tensor: Boxes with shape (N, 4), where columns represent
-            tl_x, tl_y, br_x, br_y.
+        Tensor: Boxes with shape (B, N, 4 * num_classes) or (B, N, 4) or
```
Comment (jshilong) [inline mmdet/core/bbox/coder/delta_xywh_bbox_coder.py:162]: all  `4 * num_classes` should be changed to  `num_classes * 4`

**Review by ZwwWayne — APPROVED** [review]
